include/gridtools/stencil_composition/sid/sid_helpers.hpp

```diff
@@ -0,0 +1,43 @@
+/*
+ * GridTools
+ *
+ * Copyright (c) 2014-2019, ETH Zurich
+ * All rights reserved.
+ *
+ * Please, refer to the LICENSE file in the root directory.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+#pragma once
+
+#include "../../meta.hpp"
+#include "concept.hpp"
+#include "delegate.hpp"
+#include "multi_shift.hpp"
+
+namespace gridtools {
+    namespace sid {
+        namespace sid_helpers_impl_ {
+            template <class Sid>
+            class shifted_sid : public delegate<Sid> {
+                GT_META_CALL(sid::ptr_holder_type, Sid) m_origin;
+
+                friend GT_META_CALL(sid::ptr_holder_type, Sid) sid_get_origin(shifted_sid &obj) { return obj.m_origin; }
+
+              public:
+                template <class Offsets>
+                constexpr shifted_sid(Sid &&original_sid, Offsets &&offsets) noexcept
+                    : delegate<Sid>(std::forward<Sid>(original_sid)), m_origin{[this, &offsets]() {
+                          auto &&strides = sid::get_strides(this->impl());
+                          GT_META_CALL(sid::ptr_diff_type, Sid) ptr_offset{};
+                          multi_shift(ptr_offset, strides, offsets);
+                          return sid::get_origin(this->impl()) + ptr_offset;
+                      }()} {}
+            };
+        } // namespace sid_helpers_impl_
+
+        template <class Sid, class Offset>
+        sid_helpers_impl_::shifted_sid<Sid> shifted_sid(Sid &&sid, Offset &&offset) {
```

Comment:
on naming: I would propose just `shift` because it is a good idea to name functions by verbs.
We do *shift* of the sid and we have *shifted_sid* as a result. Logical.

`_sid` suffix is redundant. we already in the `sid` namespace

Reply:
I agree on the naming, and it was originally my plan. But isn't it confusing if we have sid::shift to shift the sid and to shift the ptr/ptrdiff? Especially on the filename I don't think having `multi_shift.hpp` and `shift.hpp` next to each other.

Reply:
Good point!
How about `shift_origin` ?